Sources/Controllers/2.0/OAGPXListViewController.mm

```diff
@@ -1283,90 +994,512 @@ - (void)doSaveTrack
     }
 }
 
+- (BOOL) onSwitchClick:(id)sender
+{
+    UISwitch *sw = (UISwitch *)sender;
+    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:sw.tag & 0x3FF inSection:sw.tag >> 10];
+    NSInteger dataIndex = indexPath.row - 1;
+    OAGpxTableGroup *groupData = [_data objectAtIndex:indexPath.section];
+    NSDictionary* item = [groupData.groupItems objectAtIndex:dataIndex];
+    OAGPX *gpx = item[@"track"];
+    if (sw.isOn)
+        [_settings showGpx:@[gpx.gpxFileName] update:NO];
+    else if ([_settings.mapSettingVisibleGpx containsObject:gpx.gpxFileName])
+        [_settings hideGpx:@[gpx.gpxFileName] update:NO];
+    [self.gpxTableView reloadRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
+    [self.gpxTableView reloadSections:[NSIndexSet indexSetWithIndexesInRange:NSMakeRange(1, [self.gpxTableView numberOfSections] - 1)] withRowAnimation:UITableViewRowAnimationNone];
+    return NO;
+}
+
+
+#pragma mark - UITableViewDataSource
+
+- (NSInteger) numberOfSectionsInTableView:(UITableView *)tableView
+{
+    return _data.count;
+}
+
+- (NSString *) tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
+{
+    id sectionObject = [_data objectAtIndex:section];
+    if (section == kRecordTrackSection)
+        return OALocalizedString(@"record_trip");
+    if ([sectionObject isKindOfClass:NSArray.class])
+    {
+        NSDictionary *item = ((NSArray *)_data[section]).firstObject;
+        return item[@"header"];
+    }
+    else
+        return @"";
+}
+
+- (CGFloat) tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section
+{
+    id sectionObject = [_data objectAtIndex:section];
+    if ([sectionObject isKindOfClass:NSArray.class])
+        return 48.;
+    else if (section == 0 && [_iapHelper.trackRecording isActive])
+        return 0.01;
+    else
+        return 24.;
+}
+
+- (NSInteger) tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
+{
+    if ([[_data objectAtIndex:section] isKindOfClass:OAGpxTableGroup.class])
+    {
+        OAGpxTableGroup* groupData = [_data objectAtIndex:section];
+        if (groupData.isOpen)
+            return [groupData.groupItems count] + 1;
+        return 1;
+    }
+    return [[_data objectAtIndex:section] count];
+}
+
+- (UITableViewCell *) tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
+{
+    id sectionObject = [_data objectAtIndex:indexPath.section];
```

Comment:
Please refactor this entirely. Create an internal class(es) that will contain all the necessary section information and use them instead.